Public/Src/Cache/ContentStore/Vfs/VfsTree.cs

```diff
@@ -0,0 +1,195 @@
+    internal class VfsTree
+    {
+        private readonly ConcurrentBigMap<string, VfsNode> _nodeMap = new ConcurrentBigMap<string, VfsNode>(keyComparer: StringComparer.OrdinalIgnoreCase);
+        private readonly VfsDirectoryNode _root;
+        private readonly VfsCasConfiguration _configuration;
+
+        public VfsTree(VfsCasConfiguration configuration)
+        {
+            _configuration = configuration;
+            _root = new VfsDirectoryNode(string.Empty, DateTime.UtcNow, null);
+            _nodeMap[string.Empty] = _root;
+        }
+
+        /// <summary>
+        /// Attempts to get the VFS node (directory or file) at the given VFS root relative path
+        /// </summary>
+        public bool TryGetNode(string relativePath, out VfsNode node)
+        {
+            return _nodeMap.TryGetValue(relativePath, out node);
+        }
+
+        /// <summary>
+        /// Adds a file node (and any needed parent directory nodes) at the VFS root relative path
+        /// </summary>
+        public VfsFileNode AddFileNode(string relativePath, VfsFilePlacementData data)
+        {
+            var timestamp = DateTime.UtcNow;
+
+            if (_nodeMap.TryGetValue(relativePath, out var node))
+            {
+                return (VfsFileNode)node;
+            }
+            else
+            {
+                var parent = GetOrAddDirectoryNode(Path.GetDirectoryName(relativePath), allowAdd: true);
+                var result = _nodeMap.GetOrAdd(relativePath, (parent, timestamp, data), (l_relativePath, l_data) =>
+                {
+                    return new VfsFileNode(Path.GetFileName(l_relativePath), l_data.timestamp, l_data.parent, l_data.data.Hash, l_data.data.RealizationMode, l_data.data.AccessMode);
+                });
+
+                node = result.Item.Value;
+
+                return (VfsFileNode)node;
+            }
+        }
+
+        public VfsDirectoryNode GetOrAddDirectoryNode(string relativePath, bool allowAdd = true)
+        {
+            if (string.IsNullOrEmpty(relativePath))
+            {
+                return _root;
+            }
+
+            if (_nodeMap.TryGetValue(relativePath, out var node))
+            {
+                return (VfsDirectoryNode)node;
+            }
+            else if (allowAdd)
+            {
+                var parent = GetOrAddDirectoryNode(Path.GetDirectoryName(relativePath), allowAdd: true);
+                node = _nodeMap.GetOrAdd(relativePath, parent, (l_relativePath, l_parent) =>
+                {
+                    return new VfsDirectoryNode(Path.GetFileName(relativePath), DateTime.UtcNow, parent);
+                }).Item.Value;
+
+                return (VfsDirectoryNode)node;
+            }
+
+            return null;
+        }
+    }
+
+    public abstract class VfsNode
+    {
+        public readonly string Name;
+        public readonly DateTime Timestamp;
+        public VfsNode NextSibling;
+        public VfsNode PriorSibling;

```

Comment:
>Prior [](start = 23, length = 5)

Is this often called `PreviousSibling`?

Reply:
Not sure. They are equivalent as far as I'm concerned. I'll rename.

---
In reply to: [298340773](https://github.com/microsoft/BuildXL/pull/519#discussion_r298340773) [](ancestors = 298340773)